### ISSUE TITLE
[admin] Milestones for Wrattler beta

### ISSUE THREAD

**Issue by tpetricek** [wrattler/wrattler#9]
# Milestones

The following plan aims to build a version of Wrattler by December that works - that is, it implements roughly similar functionality as the earlier prototype, but in a more solid way that can be maintained in the long-term by the RSE team. It will let us run and play with sample data analyses (of the AIDA datasets) as well as create new ones using Python or R with one or two AI assistants (depending on what is available from the rest of the AIDA team). 

The plan is based on the assumption that we have @nbarlowATI and @myyong working on this (50% of their time each) and we also get a bit of time from @tpetricek and @triangle-man. If this changes, we will need to drop something from the plan.

### Milestone 0 (mid August)
Simple notebook running in browser with Markdown and JavaScript with basic user interface that lets us add and remove cells; in parallel, we'll have some server infrastructure for running Python and Docker orchestration for testing and demoing.

 * (Browser) TypeScript interfaces for dependency graph [M+T]
 * (Browser) Implement the above for Markdown and JavaScript [M]
 * (Browser) Recreate basic prototype user interface [M]
 * (Server) Python plugin [N] (Note, this is partly done already)
 * (Server) Docker compose to start services [N]
 * (Server) Import data store code [T]
 * No problem: Keep data store, Only client side integrations 
 
### Milestone 1 (mid September)
Notebook that lets us combine Python, JavaScript and (possibly) TheGamma in a more evolved user interface. The languages communicate via data store and can be used to create and save analyses. In parallel, we'll have analyses of AIDA datasets, ready to be integrated into this (but written separately).

 * (Server) Deploying Python plugin to Azure for demos [N]
 * (Server) Better Python plugin [N]
 * (Browser) Client-side part for Python plugin [M+N]
 * (Browser) Create decent GUI :-)
 * (Samples) Draft of challenges in Jupyter or whatever [G]
 * (Browser) Save button (!) so that Giovanni can use this [M+T]
 * (Browser) TheGamma plugin [T] (optional)
 * No problem: Keep data store

### Milestone 2 (end of November)
Sample AIDA data analyses available in the Wrattler notebook. We will add integration with Data diff and TheGamma, to have one AI assistant available and we will work on documentation and infrastructure (optionally) supporting JupyterLab and Binder integration.

 * (Documentation) How to write a plugin? [M+T]
 * (Browser) Datadiff plugin [T]
 * (Server) JupyterLab and Binder support [N+M+T]
 * (Samples) Integrating challenges from Giovanni [G+*]
 * (Documentation) www.wrattler.org

### Milestone 3 (end of February)
We will integrate second AI assistant into Wrattler. At the time of writing, this would most likely be the ptype inference assistant (which requires updates to the data store and integration work). However, this may change if other components from the AIDA team become available.

 * (Server) Revisited data store with meta-data [J+T]
 * (Server) ptype integration [J+H+T]

### Milestone X (in the future)
There is ongoing work in the AIDA team that we might put into the plan, but this requires joint work and collaboration with RAs. It would be nice to have these, but we cannot deliver this on our own.

 * Gerrit has been working on a CSV cleaner, which would make for a nice AI assistant
 * We should revisit Datadiff so that it works more interactively. See [this comment](https://github.com/alan-turing-institute/aida-datadiff/issues/36).

# Assorted notes

### What is implemented in the prototype?
 * (Browser) User interface
 * (Browser) Dependency graph 
 * (Browser) Drawing of the dependency graph
 * (Browser) Language specific plugin (R, JavaScript, TheGamma, Markdown)
 * (Server) R service
 * (Server) Data store
 * For more [see the notes on the prototype](https://github.com/wrattler/wrattler-design/blob/master/2018-04-07-prototype-notes.md)
 
### What do we need for the beta?
 * (Browser) New UI implemented in TypeScript
 * (Browser) New dependency graph implemented in TypeScript
 * (Documentation) How the dependency graph works
 * (Documentation) How Wrattler language plugins work
 * (Browser) Draw dependency graph
 * (Server) Python service (and eventually fixed R service)
 * (Server) Keep current data store 

### What are the communication protocols?
 * Protocol used by Python/R service (analyze code, run code)
 * Protocol used for talking with the data store 
 * They are both [documented here](https://github.com/wrattler/wrattler-design/blob/master/2018-04-07-prototype-notes.md)
 * We can keep those but possibly revisit them after JupyterLab integration

### More long term work items
 * (Browser) JuptyerLab integration (providing a plugin for `*.wratnb` files)
 * (Server) Binder support for starting everything
 * (Server) Data store with meta data

**Comment by tpetricek** [comment]
Following a discussion, here are two comments from @casutton about a way of looking at the milestones:

First, when will we have a version ready for internal dogfooding? The answer is that Milestone 2 is a version that can be used by insiders and brave external users and Milestone 1 is a version that can be used by brave internal users.

Second - looking at the more research aspects of the system - when will we be able to learn something about the design (that might cause us to reconsider some of it)? The answer is that the design makes very few commitments - in particular, AI assistants and Data Store can be developed fairly independently and we should be able to integrate any design. The only thing we commit to is some aspects of the dependency graph (in Milestone 0 and more in 1). 

Once we have the dependency graph, we should demonstrate that it lets us do novel things that we want to do. Examples of things we want to do are:

 * Live preview (this will be done in Milestone 1)
 * Marking stale cells (this will be done in Milestone 1)
 * Extract and translate refactoring to get clean version (this is a great research task, but we don't have enough resources to do this)
 * Tracking history of notebook (dtto.)

**Comment by tpetricek** [comment]
One more comment from Chris is that we will likely need the R service for Milestone 1, because some of the AIDA analyses will be in R.

As a least effort option, we could probably find a way to run my current implementation on some VM in Azure. A better option would be to get someone with R experience to help us clean up the current code - the main problem is figuring out how to run HTTP server that handles JSON in R.
